### PR TITLE
Integrate FxCop Analyzer

### PR DESCRIPTION
**Pull request by mark-monteiro**
Add FxCop Analyzer using the same ruleset in the main Jellyfin repository. Also includes fixes to all warnings produced by the analyzer in the existing code.

### CODE REVIEW

File: jellyfin.ruleset [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+<RuleSet Name="Rules for Emby.AutoOrganize" Description="Code analysis rules for Emby.AutoOrganize.csproj" ToolsVersion="14.0">
+  <Rules AnalyzerId="StyleCop.Analyzers" RuleNamespace="StyleCop.Analyzers">
+    <!-- disable warning CA1032: Implement standard exception constructors -->
+    <Rule Id="CA1032" Action="Info" />
+    <!-- disable warning SA1202: 'public' members must come before 'private' members -->
+    <Rule Id="SA1202" Action="Info" />
+    <!-- disable warning SA1204: Static members must appear before non-static members -->
+    <Rule Id="SA1204" Action="Info" />
+    <!-- disable warning CA2227: Collection properties should be read only -->
+    <Rule Id="CA2227" Action="Info" />
```
Comment (mark-monteiro) [inline jellyfin.ruleset]: This file is a copy of the one from the main `jellyfin` repository, with the single addition of disabling `CA2227`.
Collection properties are mostly used in DTO objects in this project, but they are not deserialized correctly when set to read-only, so I just disabled this rule.

Comment (Bond-009) [reply jellyfin.ruleset]: What changes did you try, and what broke exactly?

File: Emby.AutoOrganize/Core/EpisodeFileOrganizer.cs [hunk 2]
```diff
@@ -434,29 +427,39 @@ private NamingOptions GetNamingOptionsInternal()
...
-            var episode = await GetMatchingEpisode(series, seasonNumber, episodeNumber, endingEpiosdeNumber, result, premiereDate, cancellationToken);
+            var episode = await GetMatchingEpisode(
+                series,
+                seasonNumber,
+                episodeNumber,
+                endingEpiosdeNumber,
+                result,
+                premiereDate,
+                cancellationToken
+            ).ConfigureAwait(false);
```
Comment (Bond-009) [inline Emby.AutoOrganize/Core/EpisodeFileOrganizer.cs]: This should be at the end of the last parameter.

File: Emby.AutoOrganize/PluginEntryPoint.cs [hunk 3]
```diff
@@ -17,13 +17,13 @@
 
 namespace Emby.AutoOrganize
 {
-    public class PluginEntryPoint : IServerEntryPoint
+    public sealed class PluginEntryPoint : IServerEntryPoint
```
Comment (Bond-009) [inline Emby.AutoOrganize/PluginEntryPoint.cs:20]: Why seal this class?

File: jellyfin.ruleset (hunk 1 above)
Comment (mark-monteiro) [reply jellyfin.ruleset]: The initial steps I took to make the collections read-only were:
- Remove the setters on all collection properties
- Update code assigning values to the setters with `Clear()`/`AddRange()` where necessary
- Set default values for the collections to an empty collection (list/dictionary)

However, that seemed to break any of the web requests using those collections. Couple examples I noticed:
- In `Emby.AutoOrganize\Api\FileOrganizationService.cs`, `DeleteSmartMatchEntry.Entries` was always an empty list when received in the associated web request
- In `Emby.AutoOrganize\Model\TvFileOrganizationOptions.cs`, `LeftOverFileExtensionsToDelete` and `WatchLocations` were both always received as an empty list when updating settings

I know this type of deserialization is possible with the built in JSON serializer or with Json.NET, but I couldn't find any documentation on how to do it with ServiceStack. Any tips on how to make it work?

File: Emby.AutoOrganize/PluginEntryPoint.cs (hunk 3 above)
Comment (mark-monteiro) [reply Emby.AutoOrganize/PluginEntryPoint.cs:20]: To simplify the implementation for Dispose(). With an unsealed class you need to implement the full pattern with Dispose(bool). I can implement the full pattern if that is preferred, let me know.

File: jellyfin.ruleset (hunk 1 above)
Comment (Bond-009) [reply jellyfin.ruleset]: It’s fine, thanks for the detailed explanation.

File: Emby.AutoOrganize/PluginEntryPoint.cs (hunk 3 above)
Comment (Bond-009) [reply Emby.AutoOrganize/PluginEntryPoint.cs:20]: Ok, just wanted to hear the reasoning behind this. Since this shouldn’t ever be inherited it’s fine.

**Comment by mark-monteiro** [comment]
@Bond-009 I just copied the `jellyfin.ruleset` file directly for this PR but do we have any plans to share that file as a base ruleset across repositories? I think these instructions in the StyleCopAnalyzers documentation should work: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/Configuration.md#sharing-configuration-among-solutions

File: Emby.AutoOrganize/Core/EpisodeFileOrganizer.cs (hunk 2 above)
Comment (mark-monteiro) [reply Emby.AutoOrganize/Core/EpisodeFileOrganizer.cs]: Addressed in 77ff4a3cf7c9be911d8154059c55fea9b010d07c

**Comment by Bond-009** [comment]
@mark-monteiro that does look promising.

File: Emby.AutoOrganize/Core/EpisodeFileOrganizer.cs [hunk 4]
```diff
@@ -434,29 +427,37 @@ private NamingOptions GetNamingOptionsInternal()
...
                 episode,
                 options,
                 rememberCorrection,
                 result,
-                cancellationToken);
+                cancellationToken).ConfigureAwait(false);
         }
 
-        private async Task OrganizeEpisode(string sourcePath,
+        private Task OrganizeEpisode(string sourcePath,
```
Comment (Bond-009) [inline Emby.AutoOrganize/Core/EpisodeFileOrganizer.cs]:
```suggestion
        private Task OrganizeEpisode(
            string sourcePath,
```

File: Emby.AutoOrganize/Core/MovieFileOrganizer.cs [hunk 5]
```diff
@@ -221,7 +221,7 @@ public async Task<FileOrganizationResult> OrganizeWithCorrection(MovieFileOrgani
             FileOrganizationResult result,
             CancellationToken cancellationToken)
         {
-            var movie = GetMatchingMovie(movieName, movieYear, "", result, options);
+            var movie = GetMatchingMovie(movieName, movieYear, "", result);
```
Comment (Bond-009) [inline Emby.AutoOrganize/Core/MovieFileOrganizer.cs]:
```suggestion
            var movie = GetMatchingMovie(movieName, movieYear, string.Empty, result);
```

File: Emby.AutoOrganize/Data/BaseSqliteRepository.cs [hunk 6]
```diff
@@ -260,49 +252,31 @@ protected virtual bool EnableTempStoreMemory
...
-
         /// <summary>
         /// Releases unmanaged and - optionally - managed resources.
         /// </summary>
         /// <param name="dispose"><c>true</c> to release both managed and unmanaged resources; <c>false</c> to release only unmanaged resources.</param>
-        protected virtual void Dispose(bool dispose)
+        protected virtual void Dispose(bool disposing)
         {
-            if (dispose)
+            if (!_disposed)
```
Comment (Bond-009) [inline Emby.AutoOrganize/Data/BaseSqliteRepository.cs]: Do an early return instead

File: Emby.AutoOrganize/Data/ManagedConnection.cs [hunk 7]
```diff
@@ -7,15 +7,22 @@
 
 namespace Emby.AutoOrganize.Data
 {
-    public class ManagedConnection :  IDisposable
+    /// <summary>
+    /// Wraps a <see cref="SQLiteDatabaseConnection"/>.
+    /// </summary>
+    public sealed class ManagedConnection : IDisposable
```
Comment (Bond-009) [inline Emby.AutoOrganize/Data/ManagedConnection.cs:13]: These classes have been copied from https://github.com/jellyfin/jellyfin/tree/master/Emby.Server.Implementations/Data maybe you could reuse this documentation there ;)

File: Emby.AutoOrganize/Emby.AutoOrganize.csproj [hunk 8]
```diff
@@ -40,4 +40,14 @@
     <PackageReference Include="SQLitePCL.pretty.netstandard" Version="2.0.1" />
   </ItemGroup>
 
+  <!-- Code Analyzers-->
+  <ItemGroup Condition=" '$(Configuration)' == 'Debug' ">
+    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.7" PrivateAssets="All" />
```
Comment (Bond-009) [inline Emby.AutoOrganize/Emby.AutoOrganize.csproj]:
```suggestion
    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.8" PrivateAssets="All" />
```
Latest version ;)

File: Emby.AutoOrganize/Model/MovieFileOrganizationOptions.cs [hunk 9]
```diff
@@ -31,11 +37,11 @@ public MovieFileOrganizationOptions()
         {
             MinFileSizeMb = 50;
 
-            LeftOverFileExtensionsToDelete = new string[] { };
+            LeftOverFileExtensionsToDelete =  new List<string>();
```
Comment (Bond-009) [inline Emby.AutoOrganize/Model/MovieFileOrganizationOptions.cs]:
```suggestion
            LeftOverFileExtensionsToDelete = new List<string>();
```

File: Emby.AutoOrganize/Model/SmartMatchResult.cs [hunk 10]
```diff
@@ -5,16 +5,14 @@ namespace Emby.AutoOrganize.Model
 {
     public class SmartMatchResult
     {
-        public Guid Id { get; set; }
+        public Guid Id { get; set; } = Guid.NewGuid();
```
Comment (Bond-009) [inline Emby.AutoOrganize/Model/SmartMatchResult.cs]: ctor is preferred

File: Emby.AutoOrganize/Model/SmartMatchResult.cs [hunk 11]
```diff
@@ -5,16 +5,14 @@ namespace Emby.AutoOrganize.Model
...
+        
         public FileOrganizerType OrganizerType { get; set; }
-        public List<string> MatchStrings { get; set; }
 
-        public SmartMatchResult()
-        {
-            Id = Guid.NewGuid();
-            MatchStrings = new List<string>();
-        }
+        public List<string> MatchStrings { get; set; } = new List<string>();
```
Comment (Bond-009) [inline Emby.AutoOrganize/Model/SmartMatchResult.cs]: same

File: Emby.AutoOrganize/PluginEntryPoint.cs [hunk 12]
```diff
@@ -33,7 +33,7 @@ public class PluginEntryPoint : IServerEntryPoint
         private readonly IProviderManager _providerManager;
         private readonly IJsonSerializer _json;
 
-        public IFileOrganizationRepository Repository;
+        private IFileOrganizationRepository Repository;
```
Comment (Bond-009) [inline Emby.AutoOrganize/PluginEntryPoint.cs]: Should be lowercase and '_' prefixed

**Review by Bond-009 — CHANGES_REQUESTED** [review]

**Comment by mark-monteiro** [comment]
@Bond-009 I've addressed all of your review comments in 148d9bd.

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**